src/MDMTransaction.m

```diff
@@ -16,31 +16,108 @@
 
 #import "MDMTransaction.h"
 #import "MDMTransaction+Private.h"
+#import "MDMPerforming.h"
+#import "MDMPlan.h"
 
 @implementation MDMTransaction {
   NSMutableArray *_logs;
+  NSMutableSet *_namedPlans;
 }
 
 - (instancetype)init {
   self = [super init];
   if (self) {
     _logs = [NSMutableArray array];
+    _namedPlans = [NSMutableSet set];
   }
   return self;
 }
 
 - (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target {
-  MDMTransactionLog *log = [MDMTransactionLog new];
-  log.plans = @[ [plan copy] ];
-  log.target = target;
+  [self commonAddPlan:plan toTarget:target withName:nil];
+}
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  [self removePlanNamed:name fromTarget:target];
+  [self commonAddPlan:plan toTarget:target withName:name];
+}
+
+- (void)commonAddPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  MDMTransactionLog *log = [self newLogWithPlan:plan target:target name:name];
+  if (name.length) {
+    Class performerClass = [plan performerClass];
+    id performer = [[performerClass alloc] initWithTarget:target];
+    if ([performer conformsToProtocol:@protocol(MDMNamedPlanPerforming)] &&
+        [performer respondsToSelector:@selector(addPlan:withName:)]) {
+      [performer addPlan:plan withName:name];
+    }
+    [_namedPlans addObject:log];
+  }
   [_logs addObject:log];
 }
 
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  MDMTransactionLog *log = [self newLogWithPlan:nil target:target name:name];
+  if ([_namedPlans containsObject:log]) {
+    // need to get the log back from _namedPlans so we can get at the plans
+    MDMTransactionLog *retrievedLog = [_namedPlans member:log];
+    if ([target isEqual:retrievedLog.target]) {
+      for (id<MDMPlan>plan in [retrievedLog plans]) {
+        Class performerClass = [plan performerClass];
+        id performer = [[performerClass alloc] initWithTarget:[retrievedLog target]];
+        if ([performer conformsToProtocol:@protocol(MDMNamedPlanPerforming)] &&
+            [performer respondsToSelector:@selector(removePlan:withName:)]) {
+          [performer removePlan:plan withName:name];
+        }
+      }
+    }
+    [_namedPlans removeObject:retrievedLog];
+    [_logs removeObject:retrievedLog];
```

Comment:
Think I've hit those three points with https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/9f1c50c1de069e3bc53b08e9c9634453692f47a5, https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/06556d80d47f9a823756faedd969733451a9bd4f and https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/ca5516781ceaae56fbe212f8fe3e6d009bbf803c.

One failing test remains:
```
Test Case '-[UnitTests.SchedulerTests testPlansAreCopied]' started.
/Users/sean.oshea/github/material-motion-runtime-objc/tests/unit/SchedulerTests.swift:33: error: -[UnitTests.SchedulerTests testPlansAreCopied] : XCTAssertNotEqual failed: ("<_TtCC9UnitTests14SchedulerTestsP33_53756AB0E4D1E4CAAD92FB270B9314B713ChangeBoolean: 0x608000035fe0>") is equal to ("<_TtCC9UnitTests14SchedulerTestsP33_53756AB0E4D1E4CAAD92FB270B9314B713ChangeBoolean: 0x608000035fe0>") -
```
which is definitely related to me overriding `isEqual:` in `MDMTransactionLog`. Still looking into that.